### PR TITLE
Feature/refactor

### PR DESCRIPTION
**Pull request by omarandstuff**

### CODE REVIEW

File: app/services/sessions/retrieve.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    def initialize(token: nil, email: nil, password: nil)
+      @token, @email, @password = token, email, password
+    end
+
+    def process
+      if session.present?
+        get_session
+      else
+        create_session
+      end
```
Comment (TheNaoX) [inline app/services/sessions/retrieve.rb]: I think this class looks very good 🤓 , however I believe that the fact that we have the `get_session` and `create_session` tells me that this class has more than 1 responsibility. What do you think?

File: spec/services/sessions/retrieve_spec.rb [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+      @active_user1_session = Session.create! user: @user1, token: @active_user1_token
+      @active_user2_session = Session.create! user: @user2, token: @active_user2_token
+
+      @inactive_user1_token = Sessions::GenerateToken.new(@user1.id, 2.hours.ago).process
+      @inactive_user2_token = Sessions::GenerateToken.new(@user2.id, 2.hours.ago).process
+
+      @inactive_user1_session = Session.create! user: @user1, token: @inactive_user1_token, active: false
+      @inactive_user2_session = Session.create! user: @user2, token: @inactive_user2_token, active: false
+    end
```
Comment (TheNaoX) [inline spec/services/sessions/retrieve_spec.rb]: Did you check the article I shared with you? I think in tests we don't need to do this kind of refactorization, the more you can explain how the environment in a test is setup, the better other developers will understand what the test is about.

Comment (TheNaoX) [reply spec/services/sessions/retrieve_spec.rb]: What do you think?

**Review by TheNaoX — COMMENTED** [review]
Let me know what you think about the comments

**Comment by TheNaoX** [comment]
Also I was thinking, shall we implement a wrapper to autoexecute the services? Something like

```ruby
Sessions::GenerateToken.for(id, created_at)
```

As opposed to run

```ruby
Sessions::GenerateToken.new(id, created_at).process
```

File: spec/services/sessions/retrieve_spec.rb (hunk 2 above)
Comment (omarandstuff) [reply spec/services/sessions/retrieve_spec.rb]: Yeah I read the article, I'm always more confortable doing things the way they explane them selfs but by reading things about DRY and stuff like that, I think I feel the pressure of follow that kind of patterns but if you give me green light that would be a relief.

File: app/services/sessions/retrieve.rb (hunk 1 above)
Comment (omarandstuff) [reply app/services/sessions/retrieve.rb]: You know what, thats true, I will thing a bout what to do next with this

**Comment by omarandstuff** [comment]
I will try the wrapper, seems like a more elegant way to do it, I may need your help

File: spec/services/sessions/retrieve_spec.rb (hunk 2 above)
Comment (TheNaoX) [reply spec/services/sessions/retrieve_spec.rb]: In testing you don't need to be DRY, actually the more the test can tell you, the better, so probably you can go back to the way the tests were before the refactoring. The goal is to make tests more self-explanatory.

File: app/services/sessions/retrieve.rb (hunk 1 above)
Comment (TheNaoX) [reply app/services/sessions/retrieve.rb]: Sure, definitely you don't need to have a service for each controller, actually you can have a service for each controller action if you want, as long as it does the job, and it's SOLID

**Comment by TheNaoX** [comment]
I can give you a couple of hints for the wrapper:

* Inheritance
* A class method
* The ruby [**splat**](https://endofline.wordpress.com/2011/01/21/the-strange-ruby-splat/) operator

**Comment by TheNaoX** [comment]
I'll give you some time to think about the solution, if you need help after a couple of hours, let me know and I'll help you

File: spec/services/sessions/retrieve_spec.rb (hunk 2 above)
Comment (omarandstuff) [reply spec/services/sessions/retrieve_spec.rb]: Alright, let's do that

**Pull request merged**